distributed/scheduler.py

```diff
@@ -2014,11 +2019,29 @@ def transition_no_worker_processing(self, key, stimulus_id):
                 assert not ts.actor, f"Actors can't be in `no-worker`: {ts}"
                 assert ts in self.unrunnable
 
-            if ws := self.decide_worker_non_rootish(ts):
+            decide_worker = (
+                self.decide_worker_rootish_queuing_disabled
+                if self.is_rootish(ts)
+                else self.decide_worker_non_rootish
+            )
+            # NOTE: it's possible that queuing is enabled and `is_rootish(ts)`,
+            # meaning this task should have been queued and `decide_worker_rootish_queuing_enabled`
+            # would be the most appropriate function to use. But if, at submission time,
+            # it didn't look root-ish (`TaskGroup` too small, or cluster too big) and there were
+            # no running workers, it would have gone to `no-worker` instead.
```

Comment:
To my understanding this should be impossible.
A non-rootish task will go to `no-worker` if there are no workers available that satisfy its worker/host/resource restrictions, or if there are no workers at all.
Tasks with restrictions cannot be rootish, so this leaves us only with the second option - 0 total threads in the cluster, in which case a TaskGroup with a single task can quality as rootish and neither adding tasks to the group, nor removing them will change that.

If you add workers you can flip a task from rootish to non-rootish, but a rootish task would not be in no-worker to begin with.